### PR TITLE
Introduce ability of calling `StatusQ` methods from `NIM`

### PR DESCRIPTION
**Pull request by micieslak**
### What does the PR do

This PR introduces ability to share routines between UI and backend in a convenient way (nim backend calling cpp code).

So far calling C++ code was done always via `DOtherSide` and `NimQml`. This method has limitations as described in https://github.com/status-im/status-desktop/issues/16496

This PR introduces necessary changes to provide ability of calling C++ methods defined in `StatusQ` by both ends: within backend in `nim` code and also on the UI side, from QML or cpp.

First exemplary function `isCompressedPubKey` has been moved from nim to C++ and then this single definition is used for both nim and qml.

**Changes:**
- `StatusQ` is no longer loaded as a qt plugin in the main app and others (storybook, tests). `StatusQ` is linked as any other shared libraries. It gives following benefits:
  - allows calling exposed methods from NIM code.
  - fully solves https://github.com/status-im/status-desktop/issues/10218 - no need to import `StatusQ` before importing any other `StatusQ.XXX`, no workarounds needed. All places using that workaround has been cleaned up.
  - simplifies cmake configs and build scripts (e.g. copying `qmldir` defining the module is not needed, `STATUSQ_MODULE_PATH` removed)
  - `plugin.cpp` is removed because library is no longer intended to be loaded as a module. The type registration is moved to `typesregistration.cpp`
  - `externc.cpp` file added to expose functionalities that are intended to be called from external world via `c` interface (from `NIM` in our case)
- `statusq_bridge.nim` added as a nim counterpart of `externc.cpp`. It declares methods from `StatusQ` which can be used easily by importing `import statusq_bridge`
- calls to `statusq_registerQmlTypes()` (nim) or `registerStatusQTypes` (cpp) added to entry points of all apps. It's needed because `StatusQ` is no longer loaded as a module, triggering types registration implicitly.

> [!IMPORTANT]  
> With this PR merged, there is probably no need to add any new code to `DOtherSide` or `NimQML`. In most cases it can be done using the introduced mechanism (`externc.cpp`/`statusq_brige.nim`).

Closes: https://github.com/status-im/status-desktop/issues/16496
Fixes: https://github.com/status-im/status-desktop/issues/10218

### Architecture compliance

- [x] I am familiar with the application architecture and agreed good practices.
My PR is consistent with this document: [Status Desktop Architecture Guide](https://github.com/status-im/status-desktop/blob/master/CONTRIBUTING.md)

### CODE REVIEW

**Comment by arnetheduck** [comment]
out of interest, is there anything that prevents a nim function from being called from qml?

File: AppRun [hunk 1]
```diff
@@ -5,7 +5,7 @@ export GST_PLUGIN_PATH="${APPDIR}/usr/lib/gstreamer-1.0"
 export GST_PLUGIN_PATH_1_0="${APPDIR}/usr/lib/gstreamer-1.0"
 export GST_PLUGIN_SYSTEM_PATH="${APPDIR}/usr/lib/gstreamer-1.0"
 export GST_PLUGIN_SYSTEM_PATH_1_0="${APPDIR}/usr/lib/gstreamer-1.0"
-export LD_LIBRARY_PATH="${APPDIR}/usr/lib/:${APPDIR}/usr/lib/nss:${LD_LIBRARY_PATH}"
+export LD_LIBRARY_PATH="${APPDIR}/usr/lib/:${APPDIR}/usr/lib/nss:${APPDIR}/usr/bin:${LD_LIBRARY_PATH}"
```
Comment (caybro) [inline AppRun]: This is weird by design; no libs should go to `bin`

File: Makefile [hunk 2]
```diff
@@ -826,7 +828,7 @@ ICON_TOOL := node_modules/.bin/fileicon
 
 run-linux: nim_status_client
 	echo -e "\033[92mRunning:\033[39m bin/nim_status_client"
-	LD_LIBRARY_PATH="$(QT5_LIBDIR)":"$(STATUSGO_LIBDIR)":"$(STATUSKEYCARDGO_LIBDIR):$(LD_LIBRARY_PATH)" \
+	LD_LIBRARY_PATH="$(QT5_LIBDIR)":"$(STATUSGO_LIBDIR)":"$(STATUSKEYCARDGO_LIBDIR):$(LD_LIBRARY_PATH)":bin/StatusQ \
```
Comment (caybro) [inline Makefile:831]: Same here

**Review by caybro — APPROVED** [review]
Looks good! Nice way to transition to C++ :partying_face:

File: Makefile (hunk 2 above)
Comment (micieslak) [reply Makefile:831]: I know, it will be polished, within this draft pr I'm trying to make it working only and gather feedback if we want to proceed :)

Comment (micieslak) [reply Makefile]: finally no change there :)

**Review by jrainville — APPROVED** [review]
Looks good to me, but I know nothing

**Comment by anastasiyaig** [comment]
the app does not start on mac at all


https://github.com/user-attachments/assets/3109ed14-e87c-4309-a634-b0ed490006a5

```
Library not loaded: @rpath/libStatusQ.dylib
Referenced from: <73630B92-454F-3CD3-9D48-010FC630DD58> /Users/USER/*/StatusPR.app/Contents/MacOS/nim_status_client
Reason: tried: '/opt/homebrew/Cellar/qt@5/5.15.13_1/lib/libStatusQ.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/opt/homebrew/Cellar/qt@5/5.15.13_1/lib/libStatusQ.dylib' (no such file), '/Users/anastasiya/pull-requests/pr/StatusPR.app/Contents/MacOS/../Frameworks/libStatusQ.dylib' (no such file), '/opt/homebrew/Cellar/qt@5/5.15.13_1/lib/libStatusQ.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/opt/homebrew/Cellar/qt@5/5.15.13_1/lib/libStatusQ.dylib' (no such file), '/Users/anastasiya/pull-requests/pr/StatusPR.app/Contents/MacOS/../Frameworks/libStatusQ.dylib' (no such file)
(terminated at launch; ignore backtrace)
```

also, there is no possibility to send contact request on Linux in this PR, the address is not resolved:

image

```
ERR 2024-10-10 14:50:01.345Z error doing rpc request                    topics="rpc" tid=1848401 file=core.nim:39 methodName=ens_addressOf exception="\nstatus-go error [methodName:ens_addressOf, code:-32000, message:no contract code at given address ]\n"
```

**Comment by anastasiyaig** [comment]
building locally also does not work

```
Building: status-keycard-go
Building: QR-Code-generator
Building: resources.rcc
2094 resources added
Compiling: translations
Building: bin/nim_status_client
Custom icon assigned to file 'bin/nim_status_client' based on 'status-dev.icns'.
Running: bin/StatusDev.app/Contents/MacOS/nim_status_client
dyld[27336]: Library not loaded: @rpath/libStatusQ.dylib
  Referenced from: <6CB1DC49-4E69-3625-ACFE-F67C3E8D7D46> /Users/anastasiya/status-desktop/bin/nim_status_client
  Reason: tried: '/opt/homebrew/Cellar/qt@5/5.15.12_1/lib/libStatusQ.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/opt/homebrew/Cellar/qt@5/5.15.12_1/lib/libStatusQ.dylib' (no such file), '/Users/anastasiya/status-desktop/vendor/status-go/build/bin/libStatusQ.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/Users/anastasiya/status-desktop/vendor/status-go/build/bin/libStatusQ.dylib' (no such file), '/Users/anastasiya/status-desktop/vendor/status-keycard-go/build/libkeycard/libStatusQ.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/Users/anastasiya/status-desktop/vendor/status-keycard-go/build/libkeycard/libStatusQ.dylib' (no such file), '/opt/homebrew/Cellar/qt@5/5.15.12_1/lib/libStatusQ.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/opt/homebrew/Cellar/qt@5/5.15.12_1/lib/libStatusQ.dylib' (no such file), '/Users/anastasiya/status-desktop/vendor/status-go/build/bin/libStatusQ.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/Users/anastasiya/status-desktop/vendor/status-go/build/bin/libStatusQ.dylib' (no such file), '/Users/anastasiya/status-desktop/vendor/status-keycard-go/build/libkeycard/libStatusQ.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/Users/anastasiya/status-desktop/vendor/status-keycard-go/build/libkeycard/libStatusQ.dylib' (no such file), '/usr/local/lib/libStatusQ.dylib' (no such file), '/usr/lib/libStatusQ.dylib' (no such file, not in dyld cache)
make: *** [run-macos] Abort trap: 6
```

**Review by yakimant — APPROVED** [review]
Makefile and scripts changes look good.

File: ui/StatusQ/sandbox/sandboxapp.cpp
[the remark this answers is not in the document]
Comment (caybro) [reply ui/StatusQ/sandbox/sandboxapp.cpp:1]: Yeah... I wanted to bring this up at the meeting; I think we should just drop it :fire:

[the remark this answers is not in the document]
Comment (micieslak) [reply ui/StatusQ/sandbox/sandboxapp.cpp:1]: It's still there, but not used. Waits for some cleanup with moving some potentially useful pages to storybook.

**Comment by micieslak** [comment]
> building locally also does not work
> 
> ```
> Building: status-keycard-go
> Building: QR-Code-generator
> Building: resources.rcc
> 2094 resources added
> Compiling: translations
> Building: bin/nim_status_client
> Custom icon assigned to file 'bin/nim_status_client' based on 'status-dev.icns'.
> Running: bin/StatusDev.app/Contents/MacOS/nim_status_client
> dyld[27336]: Library not loaded: @rpath/libStatusQ.dylib
>   Referenced from: <6CB1DC49-4E69-3625-ACFE-F67C3E8D7D46> /Users/anastasiya/status-desktop/bin/nim_status_client
>   Reason: tried: '/opt/homebrew/Cellar/qt@5/5.15.12_1/lib/libStatusQ.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/opt/homebrew/Cellar/qt@5/5.15.12_1/lib/libStatusQ.dylib' (no such file), '/Users/anastasiya/status-desktop/vendor/status-go/build/bin/libStatusQ.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/Users/anastasiya/status-desktop/vendor/status-go/build/bin/libStatusQ.dylib' (no such file), '/Users/anastasiya/status-desktop/vendor/status-keycard-go/build/libkeycard/libStatusQ.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/Users/anastasiya/status-desktop/vendor/status-keycard-go/build/libkeycard/libStatusQ.dylib' (no such file), '/opt/homebrew/Cellar/qt@5/5.15.12_1/lib/libStatusQ.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/opt/homebrew/Cellar/qt@5/5.15.12_1/lib/libStatusQ.dylib' (no such file), '/Users/anastasiya/status-desktop/vendor/status-go/build/bin/libStatusQ.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/Users/anastasiya/status-desktop/vendor/status-go/build/bin/libStatusQ.dylib' (no such file), '/Users/anastasiya/status-desktop/vendor/status-keycard-go/build/libkeycard/libStatusQ.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/Users/anastasiya/status-desktop/vendor/status-keycard-go/build/libkeycard/libStatusQ.dylib' (no such file), '/usr/local/lib/libStatusQ.dylib' (no such file), '/usr/lib/libStatusQ.dylib' (no such file, not in dyld cache)
> make: *** [run-macos] Abort trap: 6
> ```

Thanks! Tomorrow I will work with @alexjba to align lib paths on mac and window (I have only linux locally).

**Review by caybro — APPROVED** [review]
Code LGTM; however I think the tricky part here will be to solve the build/install issues on macOS/Windows

File: ui/main.qml [hunk 3]
```diff
@@ -16,7 +16,7 @@ import shared.stores 1.0
 import mainui 1.0
 import AppLayouts.Onboarding 1.0
 
-import StatusQ 0.1 // Force import StatusQ plugin to load all StatusQ resources
```
[the remark this answers is not in the document]
Comment (micieslak) [reply ui/main.qml:19]: Yes, exactly, `StatusWindow` comes from this import. But comment is no longer valid.

**Comment by micieslak** [comment]
@anastasiyaig I pushed fix for the contact request problem. It was the only thing I could break regarding app functionality... and I did it :sweat: So good to have those tests.

**Comment by micieslak** [comment]
> Code LGTM; however I think the tricky part here will be to solve the build/install issues on macOS/Windows

Yeah, tomorrow I will need sb to assist me with that :)

**Comment by alexjba** [comment]
For macOs the `rpath` is configured in `config.nims`. That will fix the `Library not loaded: @rpath/libStatusQ.dylib` error. Nim can now find the dependency, but qt can't find the plugin:
`WRN 2024-10-11 07:20:35.911Z qt warning                                 topics="qt" tid=10258415 text="plugin cannot be loaded for module \"StatusQ\": '/Users/alexjbanca/Repos/status-desktop/bin/StatusQ/libStatusQ.dylib' is not a Qt plugin" file=qrc:/main.qml:19 category=default`

**Comment by saledjenic** [comment]
> Now we actually do it but indirectly, via context properties injected into qml environment. It allows to call nim method from qml, with no additional cpp code in between.

@micieslak isn't it that we're using c++ when call Nim methods from qml? That goes through the DOtherSide, so there is c++ involved.

**Comment by micieslak** [comment]
> > Now we actually do it but indirectly, via context properties injected into qml environment. It allows to call nim method from qml, with no additional cpp code in between.
> 
> @micieslak isn't it that we're using c++ when call Nim methods from qml? That goes through the DOtherSide, so there is c++ involved.

Basically this part is not touched in this pr. When we call nim from qml it goes via qt's signals slots mechanism, some dotherside code is involved but we don't need to add anything extra for every method we add.

**Comment by alexjba** [comment]
LGTM! Tested the macOS dev build + CI build and the windows dev build

**Review by alexjba — APPROVED** [review]

**Comment by micieslak** [comment]
Seems that everything is covered now:

- [x] linux run
- [x] linux packaging
- [x] mac run
- [x] mac packaging
- [x] windows run
- [x] windows packaging

**Review by osmaczko — COMMENTED** [review]
I generally support the idea of defining and exposing C++ to Nim within the same repository, as it eliminates the need for an additional PR (i.e., no NimQml). However, I’m not entirely convinced that `isCompressedPubKey` should be part of StatusQ. Ideally, StatusQ components should abstract this detail into a simple string. Additionally, other methods from the global utilities are essentially proxies to status-go (and tbh `isCompressedPubKey` should also be exposed through status-go). How should we handle this in that case?

**Review by anastasiyaig — APPROVED** [review]
I've checked the builds on all platforms, its all good

**Comment by micieslak** [comment]
> I generally support the idea of defining and exposing C++ to Nim within the same repository, as it eliminates the need for an additional PR (i.e., no NimQml). However, I’m not entirely convinced that `isCompressedPubKey` should be part of StatusQ. Ideally, StatusQ components should abstract this detail into a simple string. Additionally, other methods from the global utilities are essentially proxies to status-go (and tbh `isCompressedPubKey` should also be exposed through status-go). How should we handle this in that case?

Yes, I agree that `isCompressedPubKey`used here is probably not the luckiest example. The shift could be done in an opposite way it this case (to `status-go`). Than this method should be exposed via store, as all other backend methods. With some cleanup on the ui side it should work ok. Components like e.g. `ImportCommunityPopup` should externalize the key validation instead of doing it internally. It would reduce dependency on on the backend on the component level.

Another, long term possibility could be splitting status-go monolith into some modules which could be compiled into shared libraries separately. Then all kind of standalone, not involving networking or user data utilities could be gathered in such library. Such library could be used then from nim and also cpp, including statusq, storybook and ui-side tests. But it's quite abstract topic for now probably.

My motivation for moving `isCompressedPubKey` to cpp was removing dependency on the backend from this method. But your point is fully valid. I will think more how to arrange it better. My proposition for now is removing changes regarding `isCompressedPubKey` from this PR and leave if for further consideration. The rest of changes still beneficial, `isCompressedPubKey` is kind of addition to this pr, probably not ideal one as you noted. Wdyt?

**Review by osmaczko — APPROVED** [review]

**Pull request merged**